### PR TITLE
feat: allow get headers when decoding token

### PR DESCRIPTION
**Pull request by kpn13**
This PR in an enhancement to allow to get headers directly using JWT::decode() by allowing to pass a third param by reference.

https://github.com/firebase/php-jwt/issues/431

### CODE REVIEW

**Comment by ohader** [comment]
Extracting headers would be useful, since they might contain hints to used key identifiers.
However, instead of introducing this functionality as a side-effect of `JWT::decode(..)`, decoding the header (without verifying the signature) should be more explicit, organized in a dedicated function `JWT::decodeHeader(..)`.

**Comment by kpn13** [comment]
This is interesting, thx for your feedback. But is decoding header really a side effect? Couldn't we consider decoding a token is both decoding header and payload?

**Comment by ohader** [comment]
> This is interesting, thx for your feedback. But is decoding header really a side effect? Couldn't we consider decoding a token is both decoding header and payload?

Ah sorry, I was not clear on that. I'm referring to [Side effect (computer science)](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). It's about **how** the header is decoded. In the current patch, it's not done explicitly - e.g. decoding a JWT with an invalid signing hash would change the `$headers` by reference, but also thrown an exception (due to the wrong signature). Using a dedicated function `JWT::decodeHeader(string $jwt)` would make it that process more explicit.

File: src/JWT.php [hunk 1]
```diff
@@ -94,7 +94,8 @@ class JWT
      */
     public static function decode(
         string $jwt,
-        $keyOrKeyArray
+        $keyOrKeyArray,
+        &$headers = null
```
Comment (bshaffer) [inline src/JWT.php:98]: Let's make this typehinted
```suggestion
        stdClass &$headers = null
```

File: README.md [hunk 2]
```diff
@@ -44,9 +44,15 @@ $payload = [
  * for a list of spec-compliant algorithms.
  */
 $jwt = JWT::encode($payload, $key, 'HS256');
-$decoded = JWT::decode($jwt, new Key($key, 'HS256'));
+/**
+ * The third parameter allow you to get header values in a stdClass object (you need to instance it before).
+ * It is optional.
```
Comment (vishwarajanand) [inline README.md:49]:
```suggestion
 * Providing headers is optional. If provided with an instantiated object,
 * headers will be populated.
```

File: src/JWT.php [hunk 3]
```diff
@@ -111,6 +112,11 @@ public static function decode(
         if (null === ($header = static::jsonDecode($headerRaw))) {
             throw new UnexpectedValueException('Invalid header encoding');
         }
+
```
Comment (vishwarajanand) [inline src/JWT.php:115]:
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -44,9 +44,15 @@ $payload = [
  * for a list of spec-compliant algorithms.
  */
 $jwt = JWT::encode($payload, $key, 'HS256');
-$decoded = JWT::decode($jwt, new Key($key, 'HS256'));
+/**
+ * Providing headers is optional. If provided with an instantiated object,
+ * headers will be populated.
+ */
+$headers = new stdClass();
+$decoded = JWT::decode($jwt, new Key($key, 'HS256'), $headers);
```
Comment (vishwarajanand) [inline README.md:52]:
```suggestion
$decoded = JWT::decode($jwt, new Key($key, 'HS256'));
print_r($decoded);

// Pass a stdClass in as the third parameter to get the decoded header values
$decoded = JWT::decode($jwt, new Key($key, 'HS256'), $headers = new stdClass());
print_r($headers);
```

File: README.md [hunk 5]
```diff
@@ -44,9 +44,15 @@ $payload = [
...
-$decoded = JWT::decode($jwt, new Key($key, 'HS256'));
+/**
+ * The third parameter allow you to get header values in a stdClass object (you need to instance it before).
+ * It is optional.
+ */
+$headers = new stdClass();
+$decoded = JWT::decode($jwt, new Key($key, 'HS256'), $headers);
 
 print_r($decoded);
+print_r($headers);
```
[the remark this answers is not in the document]
Comment (vishwarajanand) [reply README.md]: added this suggestion

**Review by vishwarajanand — APPROVED** [review]
Thanks @kpn13  for working on this. LGTM
I made some comments and doc suggestions and merged them too.

**Pull request merged**